No alpine

ct-clearhaus opened the pull request:
```shell
$ docker run --rm -it -v /tmp/x.rb:/x.rb ruby:2.3-alpine ruby /x.rb
false
$ docker run --rm -it -v /tmp/x.rb:/x.rb ruby:2.3 ruby /x.rb
true
```
where `x.rb` is this:
```ruby
#!/usr/bin/ruby

require 'openssl'

def ec_key_to_pkey_public_key(ec_key)
  # EC#public_key is not a PKey public key, but an EC point.
  pub = OpenSSL::PKey::EC.new(ec_key.group)
  pub.public_key = ec_key.is_a?(OpenSSL::PKey::PKey) ? ec_key.public_key : ec_key

  pub
end

if OpenSSL::PKey::EC.new.respond_to?(:private_key?) && !OpenSSL::PKey::EC.new.respond_to?(:private?)
  class OpenSSL::PKey::EC
    def private?
      private_key?
    end
  end
end


root_key = OpenSSL::PKey::EC.new("prime256v1")
root_key.generate_key
root_public = OpenSSL::PKey::EC.new('prime256v1')
root_public.public_key = root_key.public_key

leaf_key = OpenSSL::PKey::EC.new("prime256v1")
leaf_key.generate_key
leaf_public = OpenSSL::PKey::EC.new('prime256v1')
leaf_public.public_key = leaf_key.public_key

root = OpenSSL::X509::Certificate.new
root.version = 2 # cf. RFC 5280 - to make it a "v3" certificate
root.serial = 1
root.subject = OpenSSL::X509::Name.parse "/DC=org/DC=ruby-lang/CN=Ruby CA"
root.issuer = root.subject # root CA's are "self-signed"
root.public_key = ec_key_to_pkey_public_key(root_public)
root.not_before = Time.now
root.not_after = root.not_before + 2 * 365 * 24 * 60 * 60 # 2 years validity
ef = OpenSSL::X509::ExtensionFactory.new
ef.subject_certificate = root
ef.issuer_certificate = root
root.add_extension(ef.create_extension("basicConstraints","CA:TRUE",true))
root.add_extension(ef.create_extension("keyUsage","keyCertSign, cRLSign", true))
root.add_extension(ef.create_extension("subjectKeyIdentifier","hash",false))
root.add_extension(ef.create_extension("authorityKeyIdentifier","keyid:always",false))
root.sign(root_key, OpenSSL::Digest::SHA256.new)


leaf = OpenSSL::X509::Certificate.new
leaf.version = 2
leaf.serial = 2
leaf.subject = OpenSSL::X509::Name.parse "/DC=org/DC=ruby-lang/CN=Ruby certificate"
leaf.issuer = root.subject # root CA is the issuer
leaf.public_key = ec_key_to_pkey_public_key(leaf_public)
leaf.not_before = Time.now
leaf.not_after = leaf.not_before + 1 * 365 * 24 * 60 * 60 # 1 years validity
ef = OpenSSL::X509::ExtensionFactory.new
ef.subject_certificate = leaf
ef.issuer_certificate = root
leaf.add_extension(ef.create_extension("keyUsage","digitalSignature", true))
leaf.add_extension(ef.create_extension("subjectKeyIdentifier","hash",false))
leaf.sign(root_key, OpenSSL::Digest::SHA256.new)

message = 'random message'
sig = OpenSSL::PKCS7.sign(leaf, leaf_key, message)


store = OpenSSL::X509::Store.new
store.add_cert root
puts sig.verify([leaf], store, message)
```

ct-clearhaus committed:
Use ruby:2.3 over ruby:2.3-alpine

OpenSSL signature verification does not work with the alpine version.

ct-clearhaus committed:
Check signature just after creating it

ct-clearhaus commented:
Unfortunately the docker image size changes from `268MB` to `801MB` 🙁

kse-clearhaus commented:
```
$ for v in `seq 0 8`; do echo -n "2.3.$v-alpine: "; docker run -ti --rm -v $(pwd)/x.rb:/x.rb ruby:2.3.${v}-alpine ruby /x.rb; done
2.3.0-alpine: true
2.3.1-alpine: true
2.3.2-alpine: true
2.3.3-alpine: true
2.3.4-alpine: true
2.3.5-alpine: true
2.3.6-alpine: true
2.3.7-alpine: false
2.3.8-alpine: false
```

kse-clearhaus commented:
```
$ for v in `seq 0 8`; do echo -n "2.3.$v: "; docker run -ti --rm -v $(pwd)/x.rb:/x.rb ruby:2.3.${v} ruby /x.rb; done
2.3.0: true
2.3.1: true
2.3.2: true
2.3.3: true
2.3.4: true
2.3.5: true
2.3.6: true
2.3.7: true
2.3.8: true
```

ct-clearhaus commented:
We could also just stay with `ruby:2.3.3-alpine` to follow Debian's ruby version? 🙂

kse-clearhaus commented:
I don't know what the benefit of staying with 2.3.3 would be, to be honest. But I defer to you.
If we want to stay with `2.3`, maybe we should use the `2.3-slim-stretch` image, which is only 204MB.

ct-clearhaus commented:
```bash
$ docker run --rm -it debian
root@c8ad7e357bff:/# apt-get -qqy update && apt-get -y install -y ruby 2&> /dev/null
root@c8ad7e357bff:/# ruby --version
ruby 2.3.3p222 (2016-11-21) [x86_64-linux-gnu]
```

kse-clearhaus commented:
Also, great Idea to add the signature check :+1:

kse-clearhaus reviewed (approved):
Love the check :+1: 

We could also use `ruby:2.3-slim-stretch`.

ct-clearhaus commented:
```dockerfile
FROM ruby:2.3-slim-stretch
# ...
RUN apt-get update && \
    apt-get install -y gcc libc-dev libssl-dev make ruby && \
    bundle install --without development && \
    apt-get --purge autoremove -y && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*
# ...
```
Results in `390MB`. That's not too bad.

ct-clearhaus committed:
Use ruby:2.3-slim-stretch

This gives us an image of 274MB.

ct-clearhaus commented:
@kse-clearhaus your suggestion is great! https://github.com/clearhaus/pedicel-pay/pull/11/commits/fe11f5398e57c6a3498bdd3abf0514be8c740e12

ct-clearhaus commented:
Oh well ...
```bash
$ docker run --rm -it -v /tmp/pedicel-pay:/t pedicel-pay generate-token --client-path /t --backend-path /t --amount 100
/usr/local/lib/ruby/2.3.0/digest.rb:16:in `const_missing': library not found for class Digest::SHA256 -- digest/sha2.so (LoadError)
        from /usr/local/bundle/gems/pedicel-0.0.6/lib/pedicel/ec.rb:97:in `symmetric_key'
...
```

kse-clearhaus commented:
Hmm. Interesting?

ct-clearhaus commented:
Also,
```bash
$ docker run --rm -it ruby:2.3-slim-stretch ruby --version
ruby 2.3.8p459 (2018-10-18 revision 65136) [x86_64-linux]
```

And,
```
$ docker run --rm -it -v /tmp/x.rb:/x.rb ruby:2.3-slim-stretch ruby /x.rb
true
```
so that's good, but

```
$ docker run --rm -it -v $PWD:/opt/pedicel-pay ruby:2.3-slim-stretch bash
root@24f25a444e7a:/# apt-get -qqy update
root@24f25a444e7a:/# apt-get install -y gcc make > /dev/null
debconf: delaying package configuration, since apt-utils is not installed
root@24f25a444e7a:/# cd /opt/pedicel-pay
root@24f25a444e7a:/opt/pedicel-pay# bundle install > /dev/null
root@24f25a444e7a:/opt/pedicel-pay# ./exe/pedicel-pay generate-backend --path /tmp/pp
root@24f25a444e7a:/opt/pedicel-pay# ./exe/pedicel-pay generate-client --path /tmp/pp --backend-path /tmp/pp
root@24f25a444e7a:/opt/pedicel-pay# ./exe/pedicel-pay generate-token --client-path /tmp/pp --backend-path /tmp/pp
/opt/pedicel-pay/lib/pedicel-pay/backend.rb:113:in `sign': signature is wrong (RuntimeError)
        from /opt/pedicel-pay/lib/pedicel-pay/backend.rb:59:in `encrypt_and_sign'
        from ./exe/pedicel-pay:135:in `generate_token'
        from /usr/local/bundle/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
        from /usr/local/bundle/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
        from /usr/local/bundle/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
        from /usr/local/bundle/gems/thor-0.20.3/lib/thor/base.rb:466:in `start'
        from ./exe/pedicel-pay:182:in `<main>'
```

*sigh*

ct-clearhaus committed:
Add ca and intermediate to the trust store

ct-clearhaus committed:
Use alpine, but only 2.3.6

Stuff explodes in >= 2.3.7, and I will waste no more time on this
although I would certainly like to understand it fully.

kse-clearhaus reviewed (approved):
LGTM :+1:
